src/emulation_core/mips/datapath.rs

```diff
@@ -230,37 +237,135 @@ impl MipsDatapath {
-        self.rs = (self.instruction >> 21) & 0b11111;
-        self.rt = (self.instruction >> 16) & 0b11111;
-        self.rd = (self.instruction >> 11) & 0b11111;
-        self.shamt = (self.instruction >> 6) & 0b11111;
-        self.funct = self.instruction & 0b111111;
-        self.imm = self.instruction & 0xFFFF;
+        let op: u32 = self.instruction >> 26;
+        match op {
+            0 => {
+                self.instruction_enum = Instruction::RType(RType {
+                    op: ((self.instruction >> 26) & 0x3F) as u8,
+                    rs: ((self.instruction >> 21) & 0x1F) as u8,
+                    rt: ((self.instruction >> 16) & 0x1F) as u8,
+                    rd: ((self.instruction >> 11) & 0x1F) as u8,
+                    shamt: ((self.instruction >> 6) & 0x1F) as u8,
+                    funct: (self.instruction & 0x3F) as u8,
+                })
+            }
+
+            // ORI, feels kinda scummy to put this here
+            0b001101 => {
+                self.instruction_enum = Instruction::IType(IType {
+                    op: ((self.instruction >> 26) & 0x3F) as u8,
+                    rs: ((self.instruction >> 21) & 0x1F) as u8,
+                    rt: ((self.instruction >> 16) & 0x1F) as u8,
+                    immediate: (self.instruction & 0xFFFF) as u16,
+                });
+                if let Instruction::IType(i) = self.instruction_enum {
+                    self.state.imm = i.immediate as u32;
+                }
+            }
+
+            // As far as I know, there is no distinct way differentiate an instruction outside
+            // whether they are R-type or not...This sadly this means our code design will
+            // end up less elegant than desired
+            _ => {
+                self.instruction_enum = Instruction::PlaceholderType(PlaceholderType {
+                    op: ((self.instruction >> 26) & 0x3F) as u8,
+                    rs: ((self.instruction >> 21) & 0x1F) as u8,
+                    rt: ((self.instruction >> 16) & 0x1F) as u8,
+                    rd: ((self.instruction >> 11) & 0x1F) as u8,
+                    shamt: ((self.instruction >> 6) & 0x1F) as u8,
+                    funct: (self.instruction & 0x3F) as u8,
+
+                    imm: (self.instruction & 0xFF) as u16,
+                    addr: (self.instruction & 0x03FF_FFFF) as u32,
+                })
+            }
+        }
+
+        match &self.instruction_enum {
+            Instruction::RType(r) => {
+                self.state.rs = r.rs as u32;
+                self.state.rt = r.rt as u32;
+            }
+            Instruction::IType(i) => {
+                self.state.rs = i.rs as u32;
+                self.state.rt = i.rt as u32;
+            }
+            Instruction::PlaceholderType(p) => {
+                self.state.rs = p.rs as u32;
+                self.state.rt = p.rt as u32;
+            }
+            _ => panic!("fixme"),
```

Comment:
More of a nitpick than anything else in my opinion, but I might make this a `todo!()` instead to make a distinction in code for where we need to continue work on.